# HG changeset patch
# User Rik <rik@octave.org>
# Date 1541387090 28800
#      Sun Nov 04 19:04:50 2018 -0800
# Node ID a21d7d269472fdb66d792566825ed508f14d7f2f
# Parent  700b9f046538b51175c808a7b27a87e8119c991a
struct2hdl.m: Allow restoring uiXXX objects (bug #44670).

* struct2hdl.m: Change all property names to lowercase before processing.
* struct2hdl.m (createfigure): New function to recreate figure.  Figure
is initially invisible so that loading is sped up.
* struct2hdl.m (createui): New function to recreate uiXXX objects.

diff --git a/scripts/plot/util/struct2hdl.m b/scripts/plot/util/struct2hdl.m
--- a/scripts/plot/util/struct2hdl.m
+++ b/scripts/plot/util/struct2hdl.m
@@ -91,24 +91,28 @@ function [h, pout] = struct2hdl (s, p=[]
   endif
   ## read parent (last column) in p and remove it if duplicate
   par = p(2,end);
   tst = find (p(2,:) == par);
   if (numel (tst) > 1)
     p = p(1:2, 1:(tst(end)-1));
   endif
 
+  ## Use lowercase for all properties
+  s.properties = cell2struct (struct2cell (s.properties), ...
+                              tolower (fieldnames (s.properties)));
+
   ## Place the "*mode" properties at the end to avoid having the updaters
   ## change the mode to "manual" when the value is "auto".
   names = fieldnames (s.properties);
   n = strncmp (cellfun (@fliplr, names, "uniformoutput", false), "edom", 4);
   n = (n | strcmp (names, "activepositionproperty"));
   names = [names(! n); names(n)];
-  if (strcmp (s.type, "axes"))
-    n_pos = find (strcmp (names, "position") | strcmp (names, "outerposition"));
+  n_pos = find (strcmp (names, "position") | strcmp (names, "outerposition"));
+  if (strcmp (s.type, "axes") && numel (n_pos) == 2)
     if (strcmp (s.properties.activepositionproperty, "position"))
       names{n_pos(1)} = "outerposition";
       names{n_pos(2)} = "position";
     else
       names{n_pos(1)} = "position";
       names{n_pos(2)} = "outerposition";
     endif
   endif
@@ -122,49 +126,57 @@ function [h, pout] = struct2hdl (s, p=[]
   if (strcmp (s.type, "root"))
     h = 0;
     s.properties = rmfield (s.properties, ...
                               {"callbackobject", "commandwindowsize", ...
                                "monitorpositions", "pointerwindow", ...
                                "screendepth", "screenpixelsperinch", ...
                                "screensize"});
   elseif (strcmp (s.type, "figure"))
-    h = figure ();
-    s.properties = rmfield (s.properties, ...
-                              {"currentaxes", "currentcharacter", ...
-                               "currentobject", "currentpoint", "number"});
+    [h, s] = createfigure (s);
   elseif (strcmp (s.type, "axes"))
     ## legends and colorbars are "transformed" in to normal axes
-    ## if hilev is not requested
-    if (! hilev)
+    ## if hilev is not requested.
+    if (! hilev && isfield (s.properties, "tag"))
       if (strcmp (s.properties.tag, "legend"))
         s.properties.tag = "";
         s.properties.userdata = [];
         par = gcf;
       elseif (strcmp (s.properties.tag, "colorbar"))
         s.properties.tag = "";
         s.properties.userdata = [];
         par = gcf;
       endif
     endif
-    s.properties = rmfield (s.properties, {"tightinset"});
+    if (isfield (s.properties, "tightinset"))
+      s.properties = rmfield (s.properties, {"tightinset"});
+    endif
     [h, s] = createaxes (s, p, par);
   elseif (strcmp (s.type, "line"))
     h = createline (s, par);
   elseif (strcmp (s.type, "patch"))
     [h, s] = createpatch (s, par);
   elseif (strcmp (s.type, "text"))
-    s.properties = rmfield (s.properties, "extent");
+    if (isfield (s.properties, "extent"))
+      s.properties = rmfield (s.properties, "extent");
+    endif
     h = createtext (s, par);
   elseif (strcmp (s.type, "image"))
     h = createimage (s, par);
   elseif (strcmp (s.type, "surface"))
     h = createsurface (s, par);
   elseif (strcmp (s.type, "hggroup"))
     [h, s, p] = createhg (s, p, par, hilev);
+  elseif (any (strcmp (s.type, {"uimenu", "uicontextmenu",...
+                                "uicontrol", "uipanel", "uibuttongroup",...
+                                "uitoolbar", "uipushtool"})))
+    if (isfield (s.properties, "extent"))
+      s.properties = rmfield (s.properties, "extent");
+    endif
+    [h, s] = createui (s, par);
   else
     error ("struct2hdl: %s objects are not implemented yet", s.type);
   endif
 
   ## children
   p = [p [s.handle; h]];  # [original; new]
   kids = s.children;
   nkids = length (kids);
@@ -179,19 +191,39 @@ function [h, pout] = struct2hdl (s, p=[]
 
   ## paste properties
   setprops (s, h, p, hilev);
 
   pout = p;
 
 endfunction
 
+function [h, sout] = createfigure (s)
+  ## Create figure initially invisible to speed up loading.
+  opts = {"visible", "off"};
+  if (isfield (s.properties, "integerhandle"))  # see also bug #53342.
+    opts = [opts {"integerhandle", s.properties.integerhandle}];
+    s.properties = rmfield (s.properties, "integerhandle");
+  endif
+  h = figure (opts{:});
+  rmprops = {"currentaxes", "currentcharacter", "currentobject", ...
+             "currentpoint", "number"};
+  rmprops (! isfield (s.properties, rmprops)) = [];
+  s.properties = rmfield (s.properties, rmprops);
+  if (! isfield (s.properties, "visible"))
+    s.properties.visible = "on";
+  endif
+  addmissingprops (h, s.properties);
+  sout = s;
+endfunction
+
 function [h, sout] = createaxes (s, p, par)
 
-  if (! any (strcmpi (s.properties.tag, {"colorbar", "legend"})))
+  if (! isfield (s.properties, "tag")
+      || ! any (strcmpi (s.properties.tag, {"colorbar", "legend"})))
     ## regular axes
     propval = {"position", s.properties.position};
     hid = {"__autopos_tag__", "looseinset"};
     for ii = 1:numel (hid)
       prop = hid{ii};
       if (isfield (s.properties, prop))
         val = s.properties.(prop);
         propval = [propval, prop, val];
@@ -330,16 +362,24 @@ function h = createimage (s, par)
   addmissingprops (h, s.properties);
 endfunction
 
 function h = createsurface (s, par)
   h = surface ("parent", par);
   addmissingprops (h, s.properties);
 endfunction
 
+function [h, s] = createui (s, par)
+  if (isfield (s.properties, "style") && strcmp (s.properties.style, "frame"))
+    s.type = "uipanel";  # frame is deprecated: use uipanel instead
+  endif
+  h = feval (s.type, "parent", par)
+  addmissingprops (h, s.properties);
+endfunction
+
 function [h, sout, pout] = createhg (s, p, par, hilev)
   ## Here we infer from properties the type of hggroup we should build
   ## an call corresponding high level functions
   ## We manually set "hold on" to avoid next hggroup be deleted
   ## the proper value of axes "nextplot" will finally be recovered
 
   hold on;
   if (hilev)
@@ -566,18 +606,24 @@ function [h, sout, pout] = createhg_hile
 
   sout = s;
   pout = p;
 
 endfunction
 
 function setprops (s, h, p, hilev)
 
-  if (! any (strcmpi (s.properties.tag, {"colorbar", "legend"})))
-    specs = s.children(s.special);
+  isspecial = (isfield (s.properties, "tag")
+               && any (strcmpi (s.properties.tag, {"colorbar", "legend"})));
+  if (! isspecial)
+    try
+      specs = s.children(s.special);
+    catch
+      specs = [];
+    end_try_catch
     if (isempty (specs))
       hdls = [];
     else
       hdls = [specs.handle];
     endif
     nh = length (hdls);
     msg = "";
     if (! nh)
